src/logic/BoothLogic.cpp

```diff
@@ -498,9 +500,26 @@ void BoothLogic::printerThread() {
 void BoothLogic::printMonitoringThread() {
     LOG_D(TAG, "Starting Print Monitoring Thread");
 
+    // boolean flag which indicates whether to wait for external activation of this thread
+    // from the printer thread
+    bool waitForActivation = true;
+
     while (isPrintMonitoringThreadRunning) {
-        // TODO: put this thread to sleep; awake it only when a print job has been added to the list;
-        //       stay active only as long as list of monitored print jobs is not empty
+        if (waitForActivation) {
+            // blocking wait: only accept activation from printer thread (no periodic execution of this thread)
+            LOG_D(TAG, "[Print Mon Thread] Waiting for activation from printer thread (going to sleep...)");
+            printMetricsSem.wait();

```

Comment:
probably need to use timedWait in any case in order to detect shutdown, i.e. something like:

```
while(true) {
bool timeout = sema.timedWait(5000);

if(timeout && !isPrintMonitoringThreadRunning) {
return;
}
if(stuffToDo) {
// do stuff
}
}
```

Reply:
Hi Clemens,

thanks for your review and the comments. As you've already noted, the first comments are already taken care of. I also got the printer paper orientation fixed to landscape yesterday (added screenshots in the wiki).

About the semaphore and shutdown: this would actually render the semaphore useless. It is meant to wake up the new thread when the demand to monitor print jobs is known and then keep the thread running periodically until the list of jobs has emptied, then go back to sleep and wait for another semaphore posted from the printer thread. If we always do the `timedWait`, then this is basically having the thread periodic again... and sometimes waking the thread up before the (currently) 5 seconds have vanished... which does not really add any benefit to a periodic task. Nothing really time-critical is done there - except maybe re-setting alarms because of conditiona which no longer hold true.

If we want to keep the semaphore, I'd add another `post()` of it in `BoothLogic::join()` and/or(?) `BoothLogic::stop()`. (Is BoothLogic::join()` actually ever used?)

When I used Ctrl+C to terminate self-o-mat it worked (but maybe because the thread had not been sleeping, I didn't take care).. but also threw a SEGFAULT. IIRC the SEGFAULT was there before. Does the same mechanism apply for pressing Ctrl+C and the hardware button (wired to the Arduino)? Asking because my setup currently doesn't use the HW button anymore. (Exposing the button meant users would play around with pressing buttons and terminating the application out of knowing what would happen.)

Reply:
Having the semaphore with the timeout is not the same as a periodic thread, because the periodic thread will always wake up after time X whereas the semaphore will be triggered as soon as post() is called. The periodic waiting is not a problem from a performance standpoint, I thought waking it directly is a usability choice (you mentioned the delay for error messages).


With the segfault the program is forcefully terminated, so the blocking thread does not matter. If we were to fix that segfault, then the program would not exit gracefully (because we wait for the threads to quit on shutdown).

Reply:
It's seems a little hard for me to explain. Let me try again:

The time critical part IMO is the duration (=delay) from the detection of a "printer needs attention" situation until an alarm is displayed on the screen. However, the "printer needs attention" situation is only detected while the thread has already been woken up and polling the print job states, attributes, etc periodically - because I've only found the (SELPHY) printer "throw the notification" (CUPS/IPP shows additional `job-printer-state-reasons` indicating the error) during the `processing` state and not from the very beginning directly after the job has been submitted. So the thread period would define the state/reasons polling rate and hence define the frequency at what changes can be detected - or define the delay until we can set/reset an alarm.

So it's a matter of the periodicity value (how big it is) and not if the polling shall be periodic or not - I think it is required. Unless we use some other notification mechanism from the CUPS library, I haven't checked if that's technically possible.. but may over-complicate the implementation - polling seems fine to me. Only the activation of the _print monitoring thread_ from the _printer thread_ could be explicitly signaled. After the explicit activation, the _print monitoring thread_ would check the _metrics list length_ and decide whether it keeps itself awake (and keeps polling periodically) or puts itself to sleep and waits for activation from the _print thread_ again. Keeping tracks and logging the metrics is not critical to me. But I think I would keep the period time somewhere around 5..30 seconds because otherwise users might not know why the print takes overly long.. and the "event admin" might be confused why the screen still tells out of ink/paper/tray even though they have fixed the actual printer problem already quite a while ago. That's what I meant with usability.

All in all there's no need for the semaphore. It could simply _allow_ the _print monitoring thread_ to be put to sleep (lower mean system load? less logging?). I think the deactivation of alarms will always occur in the context of the thread still having a non-empty metrics list (at least one print job being processed or pending) as the `job-printer-state-reasons` attribute is attached to a print job.

Not sure if you get the idea - otherwise, we could discuss this during 38C3. (Off topic: Have you thought about giving a lightning talk about your current project(s)? https://cfp.cccv.de/38c3-lightningtalks/cfp)

About the segfault: so it has been there before (on purpose?), I did not introduce it and it's totally expected? Is it caused by the termination signal from Ctrl+C or would it also happen when being terminated by the HW button press?